signal_test.go

```diff
@@ -0,0 +1,490 @@
+	</interface>` + introspect.IntrospectDataString + `</node> `
+)
+
+type signalServer struct {
+	conn *dbus.Conn
+	test *testing.T
+}
+
+func (server *signalServer) GetString() (string, *dbus.Error) {
+	server.test.Log("Server:", getString, "called")
+	defer server.test.Log("Server:", getString, "done")
+	return "DBUS", nil
+}
+
+func (server *signalServer) EmitSignalImmediately() *dbus.Error {
+	server.test.Log("Server:", emitSignalImmediately, "called")
+	defer server.test.Log("Server:", emitSignalImmediately, "done")
+	server.test.Log("Server: Emitting immediate", orderReceived, "signal")
+	err := server.conn.Emit(objectPath, orderReceivedFull)
+	if err != nil {
+		panic(err)
+	}
+	return nil
+}
+
+func sleep(seconds int, ch chan<- int) {
+	time.Sleep(time.Second * time.Duration(seconds))
+	if ch != nil {
+		ch <- 1
+	}
+}
+
+func (server *signalServer) EmitSignalAfterTwoSeconds() *dbus.Error {
+	server.test.Log("Server:", emitSignalAfterTwoSeconds, "called")
+	defer server.test.Log("Server:", emitSignalAfterTwoSeconds, "done")
+	go func() {
+		sleep(2, nil)
+		server.test.Log("Server: Emitting belayed", orderReceived, "signal")
+		err := server.conn.Emit(objectPath, orderReceivedFull)
+		if err != nil {
+			panic(err)
+		}
+	}()
+	return nil
+}
+
+// It is like dbus.SessionBus() but it creates new connection
+// everytime instead of reusing the already created one.
+func privateSessionBus() (conn *dbus.Conn, err error) {
+	conn, err = dbus.SessionBusPrivate()
+	if err != nil {
+		return
+	}
+	if err = conn.Auth(nil); err != nil {
+		conn.Close()
+		conn = nil
+		return
+	}
+	if err = conn.Hello(); err != nil {
+		conn.Close()
+		conn = nil
+	}
+	return
+}
+
+func getServer(test *testing.T) (*signalServer, error) {
+	conn, err := privateSessionBus()
+	if err != nil {
+		return nil, err
+	}
+	reply, err := conn.RequestName(interfaceName,
+		dbus.NameFlagDoNotQueue)
+	if err != nil {
+		return nil, err
+	}
+	if reply != dbus.RequestNameReplyPrimaryOwner {
+		return nil, errors.New("name '" + interfaceName + "' already taken")
+	}
+	server := &signalServer{conn, test}
+	err = conn.Export(server, objectPath, interfaceName)
+	if err != nil {
+		return nil, err
+	}
+	err = conn.Export(introspect.Introspectable(intro), objectPath, "org.freedesktop.DBus.Introspectable")
+	if err != nil {
+		return nil, err
+	}
+	return server, nil
+}
+
+type TestData struct {
+	conn   *dbus.Conn
+	server *signalServer
+	remote *dbus.Object
+	ch1    chan *dbus.Signal
+	ch2    chan *dbus.Signal
+	test   *testing.T
+}
+
+// Creates two separate connections to system bus, one for server, one
+// for client, creates a server and adds two channels for signal
+// watching.
+//
+// We are adding two unbufferred channels to watch signals, so we can
+// watch both of them in select, handle one of them and then be sure
+// that dbus signal handling is stuck on sending signal to other
+// channel.
+func testSetup(test *testing.T) *TestData {
```

Comment:
I empathise, but they are pretty idiomatic in go for small scopes (or for very well-established patterns like this one)